### PR TITLE
FdoSecrets: fix crash when editing settings before service is enabled

### LINKED ISSUE #4311: Crash when editing Secret-Service-Integration database groups
When I edit a database group in the secret-service-integration settings before having activated the integration, Keepass crashes.

## Expected Behavior
No crash

## Current Behavior
Crash

## Possible Solution
[NOTE]: # ( Not required, but suggest a fix/reason for the bug )

## Steps to Reproduce
1. Disable Secret-Service-Integration
2. Restart Keepass
3. Go to Settings and activate Secret-Service-Integration
4. Without clicking OK try to edit a database group
5. Crash

## Debug Info
KeePassXC - Version 2.5.3
Revision: f8c962b

Qt 5.14.1
Diagnosemodus ist deaktiviert.

Betriebssystem: Arch Linux
CPU-Architektur: x86_64
Kernel: linux 5.5.2-arch1-1

Aktivierte Erweiterungen:
- Auto-Type
- Browser-Integration
- SSH-Agent
- KeeShare (bestätigtes und unbestätigtes Teilen)
- YubiKey
- Secret-Service-Integration

Kryptographische Bibliotheken:
 libgcrypt 1.8.5

Operating system: Arch Linux btw
CPU architecture: x86_64
Kernel: 5.5.2

### PR DESCRIPTION
**Pull request by Aetf**
[TIP]:  # ( Provide a general summary of your changes in the title above ^^ )

## Type of change
[NOTE]: # ( Please remove all lines which don't apply. )
- ✅ Bug fix (non-breaking change which fixes an issue)

## Description and Context
[NOTE]: # ( Describe your changes in detail, why is this change required? )
[NOTE]: # ( Describe the context of your change. Explain large code modifications. )
[NOTE]: # ( If it fixes an open issue, please add "Fixes #XXX" as necessary )
Reworks the settings page logic so that the page is only enabled if there is indeed a service instance. Fixes the crash reported in #4311.

To avoid confusion, the page now displays a message prompting the user to save before they can edit the rest part. A new message widget is used because this prompt can appear at the same time with an existing service warning.

Fixes #4311 

## Testing strategy
[NOTE]: # ( Please describe in detail how you tested your changes. )
[TIP]:  # ( We expect new code to be covered by unit tests and documented with doc blocks! )
Tested manually the steps in #4311 no longer result in a crash.

## Checklist:
[NOTE]: # ( Please go over all the following points. )
[NOTE]: # ( Again, remove any lines which don't apply. )
[NOTE]: # ( Pull Requests that don't fulfill all [REQUIRED] requisites are likely )
[NOTE]: # ( to be sent back to you for correction or will be rejected.  )
- ✅ I have read the **CONTRIBUTING** document. **[REQUIRED]**
- ✅ My code follows the code style of this project. **[REQUIRED]**
- ✅ All new and existing tests passed. **[REQUIRED]**
- ✅ I have compiled and verified my code with `-DWITH_ASAN=ON`. **[REQUIRED]**
- ✅ My change requires a change to the documentation, and I have updated it accordingly.

### CODE REVIEW

File: src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp [hunk 1]
```diff
@@ -344,14 +337,32 @@ void SettingsWidgetFdoSecrets::checkDBusName()
...
     m_ui->warningMsg->hideMessage();
+    m_ui->enableFdoSecretService->setEnabled(true);
+}
+
+void SettingsWidgetFdoSecrets::updateServiceState()
+{
+    m_ui->tabWidget->setEnabled(m_plugin->serviceInstance() != nullptr);
+    if (m_ui->enableFdoSecretService->isChecked() && !m_plugin->serviceInstance()) {
+        m_ui->warningMsg2->showMessage(
+            "Please save the settings before you can edit the following section", MessageWidget::Information, -1);
```
Comment (droidmonkey) [inline src/fdosecrets/widgets/SettingsWidgetFdoSecrets.cpp]: Needs to be wrapped with `tr(...)`

**Comment by Aetf** [comment]
I'm also not so satisfied with the current way... But I'm not sure what's better.

> Appending a message for example?

If you mean appending the message to the existing message widget if it's visible:

The message widget is essentially a single string. And it can only show one level at a time. In this case, we can have an error/warning due to existing dbus service, and a pure informative notice about applying the change first.

If you mean appending the message to the checkbox label, or something similar to the browser integration settings page where a paragraph of text is permanently visible:

This works. But I also just feel it's not quite right.

What about a small status label right to the checkbox says something like "Integration status: running/stopped", with a mouse hover tooltip explaining that the user needs to apply the change to start the plugin?

**Comment by droidmonkey** [comment]
I like that last option. Contextual error messages are the direction we are moving across the board to make it clearer to the user.

**Comment by Aetf** [comment]
Added a tooltip when the tab area is disabled but the checkbox is checked. While at it, I also fixed the margin so it is consistent with other settings page.

image

PS.
I didn't add the status label. It looks out of place and confusing. Just the tooltip should be enough.

**Pull request merged**